neqo-common/src/incrdecoder.rs

```diff
@@ -4,162 +4,34 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-use std::convert::TryFrom;
+use std::cmp::min;
 use std::mem;
 
 use crate::codec::Decoder;
 
-/// `IncrementalDecoder` incrementally processes input from incomplete input.
-/// Use this to process data that might be incomplete when it arrives.
-///
-/// ```
-/// use neqo_common::{Decoder, IncrementalDecoder, IncrementalDecoderResult};
-///
-/// let mut dec = IncrementalDecoder::decode_vec(1);
-/// let input = &[1, 0];
-/// match dec.consume(&mut Decoder::from(&input[..])) {
-///    IncrementalDecoderResult::Buffer(v) => println!("complete {:?}", v),
-///    IncrementalDecoderResult::InProgress => println!("still waiting"),
-///    _ => unreachable!(),
-/// }
-/// ```
 #[derive(Clone, Debug)]
-pub enum IncrementalDecoder {
-    Idle,
+enum IncrementalDecoderUintState {
     BeforeVarint,
     InUint { v: u64, remaining: usize },
-    InBufferLen(Box<IncrementalDecoder>),
-    InBuffer { v: Vec<u8>, remaining: usize },
-    Ignoring { remaining: usize },
 }
 
-/// `IncrementalDecoderResult` is the result of decoding a partial input.
-/// You only get an error if the decoder completed its last instruction.
-#[derive(Debug, PartialEq)]
-pub enum IncrementalDecoderResult {
-    InProgress,
-    Uint(u64),
-    Buffer(Vec<u8>),
-    Ignored,
-    Error,
+#[derive(Clone, Debug)]
+pub struct IncrementalDecoderUint {
```

Comment:
You can collapse this and put the implementation on the enum directly.

Reply:
Or, you could make the struct simpler:

```rust
struct Blah {
   v: u64,
   remaining: usize,
}
```

This can have an automatic `Default` implementation.  If remaining is zero, then you are before any number.

Reply:
remaining==0 also mean the end of decoding. I will put it as an Option

Reply:
But remaining == 0 implies that you have already received the next value.  Isn't that OK?  Anyone using this will take the Option::Some and use it.  With remaining == 0 then they can ask for the next number without making a new object (assuming that you set v = 0 on return, that is).

Reply:
All of IncrementalDecoder* are meant to be used only once. I think I will leave it as Option<>

Reply:
That being the case, you can use `Option<NonZeroUsize>` :)